### PR TITLE
Update set-namespace function

### PR DESCRIPTION
**Pull request by Shell32-Natsu**
Change set-namespace function to kustomize based version

### CODE REVIEW

File: functions/go/set-namespace/Dockerfile
Comment (frankfarzan) [inline functions/go/set-namespace/Dockerfile:1]: What are the binary files (set-namespace-linux) being checked in?

File: functions/go/set-namespace-legacy/Dockerfile
Comment (frankfarzan) [inline functions/go/set-namespace-legacy/Dockerfile:1]: Why keep the legacy function around?

File: functions/go/set-namespace-legacy/Makefile
Comment (frankfarzan) [inline functions/go/set-namespace-legacy/Makefile:1]: We should have a top-level Makefile instead of having each function have its own copy. How's the new function being built?

File: tests/set_namespace_go.sh
Comment (frankfarzan) [inline tests/set_namespace_go.sh:52]: - We should get rid of any non-containerized tests, also generally standardize the tests as part of support for curated functions. They're pretty adhoc and hack right now.
- Should test for idempotency and other properties as discusse don the doc.

@mengqiy

File: functions/go/set-namespace/main.go
Comment (frankfarzan) [inline functions/go/set-namespace/main.go:18]: What's the --help string for this command?

The usage message should a) provide a description of this function b) Describe the function config required.

Example:

https://github.com/GoogleContainerTools/kpt-functions-sdk/blob/master/ts/hello-world/src/label_namespace.ts#L29

**Review by frankfarzan — CHANGES_REQUESTED** [review]

File: functions/go/set-namespace/Dockerfile
Comment (Shell32-Natsu) [reply functions/go/set-namespace/Dockerfile:1]: That follows the old set-namespace function. It has 2 binary files checked in, one for linux and one for mac.

File: functions/go/set-namespace-legacy/Dockerfile
Comment (Shell32-Natsu) [reply functions/go/set-namespace-legacy/Dockerfile:1]: @mengqiy

File: functions/go/set-namespace/Dockerfile
Comment (frankfarzan) [reply functions/go/set-namespace/Dockerfile:1]: There's a functions/go/set-namespace/set-namespace-linux which is being added in the new function, no? In general, we should build from source, and not check in any prebuilts in Git.

Comment (Shell32-Natsu) [reply functions/go/set-namespace/Dockerfile:1]: I guess those old binary files are checked in for exec test. Since we will avoid non-containerized tests so I will remove it.

File: tests/set_namespace_go.sh
Comment (Shell32-Natsu) [reply tests/set_namespace_go.sh]: The test files used in e2e test, (`"$SDK_REPO"/example-configs`) cannot keep idempotency. Because after applying the namespace function, there will be multiple resources with duplicate ID (group, kind, version, namespace and name tuple). I will use a subdirectory (`"$SDK_REPO"/example-configs/shipping-dev`) in the test.

File: functions/go/set-namespace/Dockerfile
Comment (mengqiy) [reply functions/go/set-namespace/Dockerfile:1]: IMO binary should not be checked in here.
And even if we want to check in some big files, GIt LFS should be used.

File: functions/go/set-namespace-legacy/Dockerfile
Comment (mengqiy) [reply functions/go/set-namespace-legacy/Dockerfile]: Initially I thought we were doing this in 2 steps:
1) we rename the old one as "legacy" and meanwhile setup the new one.
2) remove the legacy after the new one is ready (including code, test and example and README).

We also remove the old one in this PR if you want to do it.

File: functions/go/set-namespace/Dockerfile
Comment (mengqiy) [inline functions/go/set-namespace/Dockerfile:1]: Why using an older go version? 1.15 is the latest.

File: functions/go/set-namespace/README.md
Comment (mengqiy) [inline functions/go/set-namespace/README.md:1]: We should have a README, but I'm OK adding the doc and example later.

File: functions/go/set-namespace/go.mod
Comment (mengqiy) [inline functions/go/set-namespace/go.mod:3]: We should use 1.15

File: functions/go/set-namespace/go.mod
Comment (mengqiy) [inline functions/go/set-namespace/go.mod:12]: I'm a little confused here. If 0.6.4 is a released version, we don't need this line.

File: functions/go/Makefile
Comment (mengqiy) [inline functions/go/Makefile:3]: I imagine we should have a docker build target for building image: `gcr.io/kpt-functions-trusted/`+fn_name

When there are fn from external contributors, we may need to have a `curated_function` list and a `contrib_function`list. The contrib fn will go to `gcr.io/kpt-functions-contrib/`

**Review by mengqiy — COMMENTED** [review]
Thanks for helping with this.
Depending on if you want to or have cycle to write the example and README for this fn, we can add them in this PR or later. If you don't want or don't have cycle to work on the doc, I can take it over.

I agree with Frank that current tests are adhoc and hacking and we should build a testing framework and have a united way to test functions. But IMO it should not block this PR from merging.

File: functions/go/set-namespace/go.mod
Comment (Shell32-Natsu) [reply functions/go/set-namespace/go.mod:12]: We need some new changes in api module and it hasn't released. We can wait for api 0.6.5. It should be released this week or next week.

File: functions/go/Makefile
Comment (Shell32-Natsu) [reply functions/go/Makefile:3]: Currently functions in this catalog are named in this way: `gcr.io/kpt-functions/`+fn name.

@frankfarzan How do you think about this proposal.

Comment (frankfarzan) [reply functions/go/Makefile:3]: We're discussing this in @mengqiy's document, I'm advocating for a delineation of community contributed functions. I'll defer to you whether to make this change in a future PR though

Comment (frankfarzan) [inline functions/go/Makefile:3]: Add comments to variables you expect humans to edit in the future.

File: functions/go/publish-functions.sh
Comment (frankfarzan) [inline functions/go/publish-functions.sh:28]: This should probably the either way around, Makefile has a command for also building and publisher docker image. It can execute helper bash scripts as implementation detail. The GH workflows can then invoke make for everything.

This can be done in a future PR.

File: functions/go/set-namespace/main.go
Comment (frankfarzan) [inline functions/go/set-namespace/main.go:95]: by convention, 'metadata' field should come before 'data'.

File: functions/go/set-namespace/main.go
Comment (frankfarzan) [inline functions/go/set-namespace/main.go:98]: describe the default behavior. What happens when 'fieldSpecs' is not specified?

File: functions/go/set-namespace/main.go
Comment (frankfarzan) [inline functions/go/set-namespace/main.go:81]: How does handle any CRD which may be cluster-scoped?
Does it handle refs to namespaces (i.e. hardcoded mapping in Kustomize today)?

In general, the function documentation has to be very precise and comprehensive with behavior and limitations.

File: functions/go/set-namespace/Dockerfile
Comment (frankfarzan) [inline functions/go/set-namespace/Dockerfile:6]: We should standardize this base image across Go and TS functions.

File: functions/go/set-namespace/main.go
Comment (frankfarzan) [inline functions/go/set-namespace/main.go:52]: General comment about error handling, you're currently just `return err` everywhere. Have you thought about what those error messages look like, and whether it requires more context (wrapping error message)?

File: functions/go/set-namespace/main.go
Comment (frankfarzan) [inline functions/go/set-namespace/main.go:24]: Why is this the default? Why RoleBinding/ClusterRolebinding?

File: functions/go/set-namespace/main.go
Comment (frankfarzan) [inline functions/go/set-namespace/main.go:1]: where does unit tests for this logic live?

**Review by frankfarzan — CHANGES_REQUESTED** [review]

File: functions/go/set-namespace/main.go
Comment (Shell32-Natsu) [reply functions/go/set-namespace/main.go:1]: The main functions are tested in the pluginator. But I will add a test for this *function* as well.

File: functions/go/set-namespace/main.go
Comment (Shell32-Natsu) [reply functions/go/set-namespace/main.go:24]: They are from the default config in kustomize. I will remove RoleBinding/ClusterRoleBinding

File: functions/go/set-namespace/main.go
Comment (Shell32-Natsu) [reply functions/go/set-namespace/main.go:95]: This metadata is for function config. Should it be from data as wel?

File: functions/go/set-namespace/Dockerfile
Comment (mengqiy) [reply functions/go/set-namespace/Dockerfile:6]: `alpine:latest` may not be great
IMO we should use `alpine:3.12` for go functions and `node:lts-alpine3.12` for TS functions.
It's probably safe to use the latest patch releases. `alpine:3.12` points to the latest 3.12.* versions.

File: functions/go/set-namespace/go.mod
Comment (mengqiy) [reply functions/go/set-namespace/go.mod:13]: Please add a TODO comment here to remind us to do that.

File: functions/go/set-namespace/Dockerfile
Comment (Shell32-Natsu) [reply functions/go/set-namespace/Dockerfile:6]: @mengqiy It will be good to create an issue to discuss it.

File: functions/go/set-namespace/main.go
Comment (frankfarzan) [reply functions/go/set-namespace/main.go:90]: Yes, this is general convention for any KRM that GVKNN should come before the spec.

File: functions/go/set-namespace/main.go
Comment (Shell32-Natsu) [reply functions/go/set-namespace/main.go:76]: Updated the usage message.

File: functions/go/set-namespace/README.md
Comment (Shell32-Natsu) [reply functions/go/set-namespace/README.md:1]: https://github.com/GoogleContainerTools/kpt/issues/1245

File: functions/go/Makefile
Comment (Shell32-Natsu) [reply functions/go/Makefile:4]: https://github.com/GoogleContainerTools/kpt/issues/1246

File: tests/set_namespace_go.sh
Comment (frankfarzan) [inline tests/set_namespace_go.sh:54]: @mengqiy We should have all the example packages in the catalog repo. This is currently using examples in SDK repo which tightly couples the SDK and functions in the catalog.

**Review by frankfarzan — CHANGES_REQUESTED** [review]

File: functions/go/set-namespace/Dockerfile
Comment (Shell32-Natsu) [reply functions/go/set-namespace/Dockerfile:6]: https://github.com/GoogleContainerTools/kpt/issues/1247

File: functions/go/set-namespace/main_test.go
Comment (Shell32-Natsu) [reply functions/go/set-namespace/main_test.go:56]: There are some [issues](https://stackoverflow.com/a/55324723/3317502) to test main functions with unit tests in go.

File: tests/set_namespace_go.sh
Comment (Shell32-Natsu) [reply tests/set_namespace_go.sh:54]: The first 2 tests for set-namespace are using sdk repo.

File: functions/go/set-namespace/main.go
Comment (Shell32-Natsu) [reply functions/go/set-namespace/main.go:78]: There are only `namespace` and `fieldSpecs` in data field. Will add fieldSpec.

File: functions/go/set-namespace/main.go
Comment (Shell32-Natsu) [reply functions/go/set-namespace/main.go:96]: 1. `path` is a field in `fieldSpec` and it is reused by multiple different transformers.
2. Is it impossible that user's CRDs won't have namespace in different paths?

File: functions/go/set-namespace/main.go
Comment (Shell32-Natsu) [reply functions/go/set-namespace/main.go]: I don't get your point. What do you mean 'what the native k8s resources are covered'? Are you talking about the API schema version?

Comment (frankfarzan) [reply functions/go/set-namespace/main.go]: "This information is collected from Kubernetes openAPI schema"

Currently, when you run this function with kpt, which resources are covered in the OpenAPI spec?

File: functions/go/set-namespace/main_test.go
Comment (frankfarzan) [reply functions/go/set-namespace/main_test.go:56]: To clarify, my point is that this should either be a:
- e2e tests which invokes the function using KPT CLI
- unit tests for underlying go package that implements the function

main_test.go is kind of a weird hybrid.

File: tests/set_namespace_go.sh
Comment (frankfarzan) [reply tests/set_namespace_go.sh:54]: They shouldn't. We should have all the example packages checked in the catalog repo. Otherwise, a PR in SDK repo can break tests here.

File: functions/go/publish-functions.sh
Comment (frankfarzan) [reply functions/go/publish-functions.sh:29]: @mengqiy @Shell32-Natsu Probably want to add a name to the TODO or track all these issues in a bug. Anti pattern would be adding a TODOs that are not addressed.

File: functions/go/set-namespace/main.go
Comment (frankfarzan) [reply functions/go/set-namespace/main.go:17]: How was this addressed?

File: functions/go/set-namespace/main.go
Comment (frankfarzan) [reply functions/go/set-namespace/main.go:98]: I think this would be the too low-level an abstraction to provide to the user.  Otherwise, we can just provide a generic function called `transformer` that has two fields:

1. fieldSpec
2. Value to set on fields matched by fieldSpec

This is similar to Search and Replace functionality. 

Instead, a user should just able to use a function without any configuration. The function should be smart enough to change a namespace across resource. In other words, the complexity should be taken care by the function instead of pushed as a burden to end users. We're not there yet, this requires two things:
1. Ability to configure the OpenAPI document used by kpt (design in progress)
2. OpenAPI document capturing references to namespaces.

Comment (Shell32-Natsu) [reply functions/go/set-namespace/main.go:98]: User can use the function without configuration. That's why we have a default fieldSpec. We can get rid of fieldSpec after the OpenAPI implementation. If there is no `path`, user cannot customize the function.

File: functions/go/set-namespace/main.go
Comment (Shell32-Natsu) [reply functions/go/set-namespace/main.go:17]: `usage` function provides the --help message.

File: tests/set_namespace_go.sh
Comment (Shell32-Natsu) [reply tests/set_namespace_go.sh:54]: https://github.com/GoogleContainerTools/kpt/issues/1248

File: functions/go/set-namespace/main.go
Comment (Shell32-Natsu) [reply functions/go/set-namespace/main.go]: Added more details about API schema.

File: functions/go/set-namespace/main_test.go
Comment (Shell32-Natsu) [reply functions/go/set-namespace/main_test.go]: Removed main_test.go and add NamespaceTransformer_test.go

File: functions/go/set-namespace/Dockerfile
Comment (mengqiy) [reply functions/go/set-namespace/Dockerfile:6]: I have added a section in my doc and you guys should get an email notification for that. PTAL.

File: functions/go/publish-functions.sh
Comment (frankfarzan) [reply functions/go/publish-functions.sh:29]: Ping. Who's following up on this? Is that a follow up PR?

File: functions/go/set-namespace/main.go
Comment (frankfarzan) [reply functions/go/set-namespace/main.go:94]: I think we're in agreement this this is a crutch until the function itself can automatically handle the use case.

@mengqiy We may want to capture this in a backlog issue and add TODOs here.

**Review by frankfarzan — APPROVED** [review]
approved once few comments are addressed. Please wait for @mengqiy as well.

**Review by mengqiy — APPROVED** [review]
I'm going to file issues to track the open issues shortly.

File: functions/go/set-namespace/main.go
Comment (mengqiy) [reply functions/go/set-namespace/main.go:94]: Filed https://github.com/GoogleContainerTools/kpt/issues/1252.

File: functions/go/publish-functions.sh
Comment (mengqiy) [reply functions/go/publish-functions.sh:29]: filed https://github.com/GoogleContainerTools/kpt/issues/1251

**Pull request merged**
